packages/nextjs/pages/blockexplorer.tsx

```diff
@@ -0,0 +1,172 @@
+import React, { useEffect, useState } from "react";
+import { useRouter } from "next/router";
+import { BlockWithTransactions, TransactionResponse } from "@ethersproject/abstract-provider";
+import { ethers } from "ethers";
+import type { NextPage } from "next";
+import { ArrowLeftIcon, ArrowRightIcon } from "@heroicons/react/24/outline";
+import { SearchBar } from "~~/components/blockexplorer/SearchBar";
+import { TransactionsTable } from "~~/components/blockexplorer/TransactionsTable";
+import deployedContracts from "~~/generated/deployedContracts";
+import { getTargetNetwork } from "~~/utils/scaffold-eth";
+
+// TODO: Add OPCODES and maybe STORAGE for contracts
+// TODO: Better pagination(some pages have less than BLOCKS_PER_PAGE blocks)
+// TODO: Refactor, seperate the table and the search bar to a component
+// TODO: Don't do everything in a loop... Please...
+// TODO: try useMemo to avoid unnecessary re-renders
+// TODO: Error handling for API calls
+// TODO: Add tests
+
+type TransactionWithFunction = TransactionResponse & { functionName?: string };
+
+const BLOCKS_PER_PAGE = 20;
+
+const Blockexplorer: NextPage = () => {
+  const [blocks, setBlocks] = useState<BlockWithTransactions[]>([]);
+  const [transactionReceipts, setTransactionReceipts] = useState<{
+    [key: string]: ethers.providers.TransactionReceipt;
+  }>({});
+  const [currentPage, setCurrentPage] = useState(0);
+  const [totalBlocks, setTotalBlocks] = useState(0);
+  const [searchInput, setSearchInput] = useState("");
+
+  const configuredNetwork = getTargetNetwork();
+  const router = useRouter();
+
+  const provider = new ethers.providers.JsonRpcProvider("http://localhost:8545");
+
+  const fetchBlocks = async () => {
+    const blockNumber = await provider.getBlockNumber();
+    setTotalBlocks(blockNumber);
+    const blocks: BlockWithTransactions[] = [];
+    const receipts: { [key: string]: ethers.providers.TransactionReceipt } = {};
+    const startingBlock = blockNumber - currentPage * BLOCKS_PER_PAGE;
+
+    for (let i = 0; i < BLOCKS_PER_PAGE; i++) {
+      const blockNumberToFetch = startingBlock - i;
+      if (blockNumberToFetch < 0) {
+        break;
+      }
+
+      const block = (await provider.getBlockWithTransactions(blockNumberToFetch)) as BlockWithTransactions & {
```

Comment:
We should avoid `as` assertion and find better way since `getBlockWithTransaction` transactions clearly won't have `functionName`property although we are dynamically asssiging key in below

Please ignore this comment, just leavning note here so that we can fix this before merging